### PR TITLE
Major simplification of PETSc and SLEPc detection

### PR DESCRIPTION
**Pull request by garth-wells**
Use CMake to automatically process the PETSc and SLEPc pkg-config files and generate targets.

### CODE REVIEW

**Comment by francesco-ballarin** [comment]
Heads up to @drew-parsons that this PR raises the minimum required PETSc/SLEPc to 3.15, while current versions on impish are 3.14, so that he is aware of this before his next build of the PPA.

**Comment by garth-wells** [comment]
We could maybe drop to 3.14, but it's completely untested.

**Comment by drew-parsons** [comment]
Thanks for the notice.  The next oldest supported Ubuntu release is focal with PETSc 3.12.  Do we know definitely that current dolfinx (if adjusted against the 3.15 detection in this PR) will not build against PETSc 3.12?

As for testing PETSc 3.14, we can consider the impish build to be the test. If it fails, we can revert the PPA and leave impish on dolfinx 0.3.0.

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -168,13 +169,29 @@ if("XTENSOR_USE_XSIMD" IN_LIST BASIX_DEFN)
...
 find_package(xtensor 0.23.10 REQUIRED)
 set_package_properties(xtensor PROPERTIES TYPE REQUIRED
   DESCRIPTION "C++ library for numerical analysis with multi-dimensional array expressions."
   URL "https://xtensor.readthedocs.io/")
 
-find_package(PETSc 3.10)
+# Check for PETSc
+find_package(PkgConfig REQUIRED)
+set(ENV{PKG_CONFIG_PATH} "$ENV{PETSC_DIR}/$ENV{PETSC_ARCH}/lib/pkgconfig:$ENV{PETSC_DIR}/lib/pkgconfig:$ENV{PKG_CONFIG_PATH}")
+pkg_search_module(PETSC REQUIRED IMPORTED_TARGET PETSc>=3.15 petsc>=3.15)
```
Comment (drew-parsons) [inline cpp/CMakeLists.txt:181]: Closes #1612

**Pull request merged**